### PR TITLE
[BEAM-1223] Introduced Sum,Min,Max#ofLongs()/ofDoubles()/ofIntegers()…

### PR DESCRIPTION
**Pull request by staslev**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

R: @lukecwik 

About half way into this I was questioning the benefit-cost ratio, but since consistency is what this PR is all about I figured bailing out would not be appropriate. 

See also the [dev list thread](https://lists.apache.org/thread.html/5d8e905ee49b116d13811c2a96da65eeb44ab7c002870f50936ee1ad@%3Cdev.beam.apache.org%3E) that discussed this.

Let me know how it looks.

### CODE REVIEW

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkRuntimeContext.java [hunk 1]
```diff
@@ -114,35 +118,6 @@ public CoderRegistry getCoderRegistry() {
     return coderRegistry;
   }
 
-  private Coder<?> getCoder(Combine.CombineFn<?, ?, ?> combiner) {
```
Comment (staslev) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkRuntimeContext.java]: R: @amitsela

Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkRuntimeContext.java]: LGTM - this change seems to keep the same functionality with far less lines of code so 👍 👍.

**Comment by staslev** [comment]
I keep getting this exception in the above tests:
```java
java.lang.IllegalAccessError: tried to access class org.apache.beam.sdk.transforms.Sum$SumIntegerFn from class com.google.cloud.dataflow.worker.counters.CounterAggregator
```

I guess it's related to deserialising `Sum$SumIntegerFn` (and other `Sum$Sum[*]Fn`s) which is a `private static class` (but has a valid no-arg ctor). 
It seems to be happening in `com.google.cloud.dataflow.worker.counters.CounterAggregator` so I don't really have access to the source code. 

@lukecwik , any words of wisdom would be much appreciated!

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java [hunk 2]
```diff
@@ -110,13 +110,69 @@ private Min() {
...
+   *
+   * @param <T> the type of the values being compared
+   */
+  public static <T, ComparatorT extends Comparator<? super T> & Serializable>
+  BinaryCombineFn<T> of(T identity, ComparatorT comparator) {
+    return new MinFn<T>(identity, comparator);
+  }
+
+  /**
+   * A {@code CombineFn} that computes the maximum of a collection of elements of type {@code T}
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java]: Fix typo from original comment `maximum -> minimum`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java [hunk 3]
```diff
@@ -110,13 +110,69 @@ private Min() {
...
+  /**
+   * A {@code CombineFn} that computes the minimum of a collection of {@code Double}s, useful as an
+   * argument to {@link Combine#globally} or {@link Combine#perKey}.
+   */
+  public static Combine.BinaryCombineDoubleFn ofDoubles() {
+    return new Min.MinDoubleFn();
+  }
+
+  /**
+   * A {@code CombineFn} that computes the maximum of a collection of elements of type {@code T}
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java]: Fix typo from original comment `maximum -> minimum`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Mean.java [hunk 4]
```diff
@@ -97,13 +94,19 @@ private Mean() { } // Namespace only
    *
    * @param <NumT> the type of the {@code Number}s being combined
    */
-  static class MeanFn<NumT extends Number>
+  public static <NumT extends Number>
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Mean.java]: Why did you need to make this public?

Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Mean.java]: Isn't `MeanFn` a type of `AccumulatingCombineFn` users might want to use?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/aggregators/NamedAggregators.java [hunk 5]
```diff
@@ -153,7 +153,9 @@ public String toString() {
   }
 
   /**
-   * =&gt; combineFunction in data flow.
+   * @param <InputT>    Input data type
```
Comment (staslev) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/aggregators/NamedAggregators.java]: 👍

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java (hunk 2 above)
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java]: 👍

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java (hunk 3 above)
Comment (staslev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java]: 👍

**Comment by staslev** [comment]
@lukecwik thanks for taking a look.

Sounds good, I'll leave the `Sum*Fn`, `Min*Fn`, `Max*Fn` and `MeanFn` public for now and create a follow up JIRA.

**Comment by lukecwik** [comment]
Instead of a JIRA, you can create a follow up PR that makes them private separate from this one and I'll merge that once I get the Dataflow worker updated.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Mean.java (hunk 4 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Mean.java]: SGTM

**Comment by staslev** [comment]
Cool, even better.

I went over your comments, let me know if I missed anything in this PR, and once we're good I'll create a new one based on this changeset but with the visibility updates in.

**Comment by lukecwik** [comment]
Yeah, checkstyle is failing:
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Mean.java:104: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Max.java:215: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Max.java:244: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Max.java:257: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Max.java:270: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Sum.java:143: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Sum.java:156: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Sum.java:169: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java:215: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java:244: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java:257: Missing a Javadoc comment. [JavadocType]
[ERROR] /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_MavenInstall/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Min.java:270: Missing a Javadoc comment. [JavadocType]

Looks like we need to keep the class comments around since they are public, just keep the existing ones that you copied over to the static factory methods.

**Comment by lukecwik** [comment]
Once the checks go green, I can merge this in.

**Comment by staslev** [comment]
Fixed conflicts with #1718 and squashed.

**Comment by lukecwik** [comment]
LGTM
Merged
